sched/sched/sched_note.c

```diff
@@ -100,6 +127,175 @@ static void note_common(FAR struct tcb_s *tcb,
   note->nc_systime[3] = (uint8_t)((systime >> 24) & 0xff);
 }
 
+/****************************************************************************
+ * Name: note_isenabled
+ *
+ * Description:
+ *   Check whether the instrumentation is enabled.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   True is returned if the instrumentation is enabled.
+ *
+ ****************************************************************************/
+
+static inline int note_isenabled(void)
+{
+#ifdef CONFIG_SCHED_INSTRUMENTATION_FILTER
+  if (!g_note_filter.mode.enable)
+    {
+      return false;
+    }
+
+#ifdef CONFIG_SMP
+  /* Ignore notes that are not in the set of monitored CPUs */
+
+  if ((g_note_filter.mode.cpuset & (1 << this_cpu())) == 0)
+    {
+      /* Not in the set of monitored CPUs.  Do not log the note. */
+
+      return false;
+    }
+#endif
+#endif
+
+  return true;
+}
+
+/****************************************************************************
+ * Name: note_isenabled_syscall
+ *
+ * Description:
+ *   Check whether the syscall instrumentation is enabled.
+ *
+ * Input Parameters:
+ *   nr    - syscall number
+ *   tcb   - The TCB of the thread
+ *   enter - syscall enter/leave flag
+ *
+ * Returned Value:
+ *   True is returned if the instrumentation is enabled.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+static inline int note_isenabled_syscall(int nr,
+                                         FAR struct tcb_s *tcb, bool enter)
+{
+#ifdef CONFIG_SCHED_INSTRUMENTATION_FILTER
+  irqstate_t irq_mask;
+#endif
+
+  if (!note_isenabled())
+    {
+      return false;
+    }
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_FILTER
+  /* sched_note_syscall_enter() and sched_note_syscall_leave() will be
+   * called from not only applications, but interrupt handlers and many
+   * kernel APIs. Exclude such situations.
+   */
+
+  if (up_interrupt_context())
```

Comment:
The same reason to https://github.com/apache/incubator-nuttx/pull/1377/files#r498625054
tracecompass cannot handle the syscall in interrupt handler correctly.

So I'll take the similar solution to nested syscalls. Remove the syscall inside the interrupt handler by app code.

But I'll not record the syscall when:
- The syscall is issued in the interrupt handler
- But the handling IRQ is not traced

Because in such case we cannot distinguish whether the syscall is called in the app code or interrupt handler.

Reply:
> - But the handling IRQ is not traced

@xiaoxiang781216 It seems that there are no way to get the IRQ number currently executing inside the interrupt handler. 
sched_note_syscall_enter/leave() cannot decide whether the syscall should be recorded or not in the interrupt handler.
So we have to leave this as it is.

Reply:
> The same reason to https://github.com/apache/incubator-nuttx/pull/1377/files#r498625054
> tracecompass cannot handle the syscall in interrupt handler correctly.
> 
> So I'll take the similar solution to nested syscalls. Remove the syscall inside the interrupt handler by app code.
> 
> But I'll not record the syscall when:
> 
> * The syscall is issued in the interrupt handler
> * But the handling IRQ is not traced
> 
> Because in such case we cannot distinguish whether the syscall is called in the app code or interrupt handler.

But if we turn off IRQ trace in the first place, how tracecompass detect whether syscall happen inside irq context or not?

> > * But the handling IRQ is not traced
> 
> @xiaoxiang781216 It seems that there are no way to get the IRQ number currently executing inside the interrupt handler.
> sched_note_syscall_enter/leave() cannot decide whether the syscall should be recorded or not in the interrupt handler.
> So we have to leave this as it is.

The post process tool could maintain a count which:
1. Increase the count in the record of sched_note_irq_enter
2. Decrease the count in the reord of sched_note_irq_leave
3. Filter out the syscall if the count isn't equal to zero
Yes, the user can disable all or partial irq trace, but tracecompass should handle the syscall record correctly if there isn't irq related record between them, regardless wether it's due to either no irq happen at all or irq is filtered by user.

Reply:
> But if we turn off IRQ trace in the first place, how tracecompass detect whether syscall happen inside irq context or not?

If we turn off all IRQ trace, all activities while executing the interrupt handler should not be recorded as note. We can realize it by simply fixing here to the followings:
```
  if (up_interrupt_context() && !(g_note_filter.mode.flag & NOTE_FILTER_MODE_FLAG_IRQ))
    {
      return false;
    }
```

And if we turn on all IRQ trace, the system call called in the interrupt handler should be recoded.
Because note_irqhandler is also recorded, the app code can maintain a count to remove the syscall as you mentioned.

My concern is when we turn off IRQ trace partially. Of course sched_note_irqhandler() knows the IRQ number and it can decide not to record the note if the IRQ is disabled to trace. But sched_note_syscall_enter/leave() cannot decide whether the note should be recorded.
So I found the idea for it, to add a static variable (or NCPUS array when SMP) in sched_note.c to keep the flag whether the running context is interrupt handler and the IRQ is disabled to trace. It is set by note_isenabled_irq() and referred by note_isenabled_syscall().

Reply:
I am wondering why we must couple syscall and irq trace. Can we only check syscall flag/mask in note_isenabled_syscall and irq flag/mask in note_isenabled_irqhandler? If the tool(e.g. tracecompass) can't handle the complex case correctly, the tool in userspace can filter out the combination with the command line option.

Reply:
For example, when an interrupt handler is called while some task is running and it issues a syscall (sem_post), the trace log would be like the followings:
```
    test-5   [0]  30.000000000: sys_write()
    test-5   [0]  30.010000000: sys_write -> 0x10
    test-5   [0]  30.030000000: irq_handler_entry: irq=15
    test-5   [0]  30.040000000: sys_sem_post()
    test-5   [0]  30.050000000: sys_sem_post -> 0x0
    test-5   [0]  30.060000000: irq_handler_exit: irq=15
    test-5   [0]  30.100000000: sys_write()
    test-5   [0]  30.110000000: sys_write -> 0x20
```
If IRQ 15 is masked and only note_isenabled_irqhandler() checks it, the log becomes:
```
    test-5   [0]  30.000000000: sys_write()
    test-5   [0]  30.010000000: sys_write -> 0x10
    test-5   [0]  30.040000000: sys_sem_post()
    test-5   [0]  30.050000000: sys_sem_post -> 0x0
    test-5   [0]  30.100000000: sys_write()
    test-5   [0]  30.110000000: sys_write -> 0x20
```
In spite of the task ID 5 only issues write(), the log seems the task also issues sem_post(). It is very confusing.
So we should couple syscalls which is called by interrupt handler and its IRQ trace. Then we'll get the following result:
```
    test-5   [0]  30.000000000: sys_write()
    test-5   [0]  30.010000000: sys_write -> 0x10
    test-5   [0]  30.100000000: sys_write()
    test-5   [0]  30.110000000: sys_write -> 0x20
```
If we don't trace IRQ, this is correct from the point of view of the task ID 5.

Reply:
@YuuichiNakamura thanks for explanation, all make sense.